c2rust-analyze/src/borrowck/mod.rs

```diff
@@ -204,11 +202,30 @@ fn run_polonius<'tcx>(
     let mut facts = AllFacts::default();
     let mut maps = AtomMaps::default();
 
-    // Start the origin counter at 3.  This has no effect on the semantics, but makes for easier
-    // diffs between our facts and the facts generated by rustc.
-    for _ in 0..3 {
-        let _ = maps.origin();
-    }
+    // the 'static region is always 0
+    let static_origin = maps.origin();
+
+    // polonius gives an origin to each generic lifetime argument
+    let mut func_lifetime_origins = vec![];
+    mir.source.def_id().as_local().map(|local_def_id| {
+        let hir_func_item = acx
+            .tcx()
+            .hir()
+            .find(acx.tcx().hir().local_def_id_to_hir_id(local_def_id));
+
+        match hir_func_item {
+            Some(Node::Item(item)) => {
+                if let ItemKind::Fn(_, ref generics, _) = item.kind {
+                    for generic in generics.params.iter() {
+                        if matches!(generic.kind, GenericParamKind::Lifetime { .. }) {
+                            func_lifetime_origins.push(maps.origin())
```

Comment:
There are a bunch of relations that rustc would normally add over these origins - please either emit the relevant facts or add comments pointing out that they're missing:

* My read of the [polonius docs](https://rust-lang.github.io/polonius/rules/relations.html#3-placeholder-and-universal_region) is that we should emit `universal_region(origin)` and `placeholder(origin, loan)` (where `loan` is a fresh loan, not connected to anything else) for each of these region parameters and also for `'static` (which is treated like a parameter).
* There should be either `known_placeholder_subset` or `subset_base` relations matching each outlives constraint `'a: 'b` in `tcx.predicates_of(local_def_id)`.
* There should be similar subset relations for the implicit outlives constraints `'static: 'a` for each parameter `'a`.

Reply:
Does this match what you suggest? d6220b75

Reply:
It seems like there is one remaining known placeholder subset relationship between lifetimes life `arg: &'a &'c i32`. I can look into what it would take to add it to this PR or we can save it for another PR

Reply:
That looks about right to me.

The `'c: 'a` bound that's implied by well-formedness of `arg`'s type should definitely show up somewhere.  I think I've seen bounds like that in the output of either `predicates_of` or a related query.  Maybe try looking at `explicit_predicates_of`, `inferred_outlives_of`, and/or `predicates_defined_on`.

It would also be fine to just leave a comment about the limitation, especially if you can't find an easy way to get the predicate from rustc.

Reply:
`predicates_defined_on` seems to encapsulate the former two, and `predicates_of` is supposed to almost-always be exactly `predicates_defined_on` (according to the documentation), but I don't see that any of these options capture the implicit relationship. i'll keep digging

Reply:
I've created an issue for this: https://github.com/immunant/c2rust/issues/1002#issue-1816492960